File: scala/core/src/main/scala/blockchain/ledger/TransactionOutputState.scala
package blockchain.ledger

import blockchain.*
import blockchain.codecs.given
import blockchain.models.*
import cats.data.OptionT
import cats.effect.{Async, MonadCancelThrow, Resource}
import cats.implicits.*
import scodec.bits.BitVector

trait TransactionOutputState[F[_]]:
  def transactionOutputIsSpendable(blockId: BlockId, outputReference: TransactionOutputReference): F[Boolean]

object TransactionOutputState:
  type State[F[_]] = Store[F, TransactionId, BitVector]
  type BSS[F[_]] = BlockSourcedState[F, State[F]]

  def make[F[_]: MonadCancelThrow](bss: BSS[F]): Resource[F, TransactionOutputState[F]] =
    Resource.pure(new TransactionOutputStateImpl[F](bss))

  def makeBSS[F[_]: Async](
      initialState: F[State[F]],
      initialBlockId: F[BlockId],
      blockIdTree: BlockIdTree[F],
      onBlockChanged: BlockId => F[Unit],
      fetchBody: FetchBody[F],
      fetchTransaction: FetchTransaction[F]
  ): Resource[F, BSS[F]] =
    new TransactionOutputStateBSSImpl[F](fetchBody, fetchTransaction).makeBss(
      initialState,
      initialBlockId,
      blockIdTree,
      onBlockChanged
    )

class TransactionOutputStateImpl[F[_]: MonadCancelThrow](bss: TransactionOutputState.BSS[F])
    extends TransactionOutputState[F]:
  override def transactionOutputIsSpendable(
      blockId: BlockId,
      outputReference: TransactionOutputReference
  ): F[Boolean] =
    OptionT
      .fromOption(outputReference.transactionId)
      .getOrRaise(new IllegalArgumentException("Illegal SelfSpend"))
      .flatMap(transactionId =>
        OptionT(bss.stateAt(blockId).use(_.get(transactionId)))
          .subflatMap(_.lift(outputReference.index))
          .exists(identity)
      )

class TransactionOutputStateBSSImpl[F[_]: Async](fetchBody: FetchBody[F], fetchTransaction: FetchTransaction[F]):
  def makeBss(
      initialState: F[TransactionOutputState.State[F]],
      initialBlockId: F[BlockId],
      blockIdTree: BlockIdTree[F],
      onBlockChanged: BlockId => F[Unit]
  ): Resource[F, TransactionOutputState.BSS[F]] =
    BlockSourcedState.make[F, TransactionOutputState.State[F]](
      initialState,
      initialBlockId,
      applyBlock,
      unapplyBlock,
      blockIdTree,
      onBlockChanged
    )

  def applyBlock(
      state: TransactionOutputState.State[F],
      blockId: BlockId
  ): F[TransactionOutputState.State[F]] =
    fetchBody(blockId)
      .map(_.transactionIds)
      .flatMap(
        _.foldLeftM(state)((state, transactionId) =>
          fetchTransaction(transactionId)
            .flatMap(transaction =>
              transaction.inputs.foldLeftM(state)((state, input) =>
                OptionT
                  .fromOption[F](input.reference.transactionId)
                  .getOrRaise(new IllegalArgumentException("Illegal SelfSpend"))
                  .flatMap(transactionId =>
                    state
                      .getOrRaise(transactionId)
                      .map(v => v.clear(input.reference.index))
                      .flatMap(updated =>
                        if (updated.populationCount > 1)
                          state.put(transactionId, updated)
                        else state.remove(transactionId)
                      )
                  )
                  .as(state)
              ) *>
                state.put(transactionId, BitVector.high(transaction.outputs.length))
            )
            .as(state)
        )
      )
      .as(state)
  def unapplyBlock(
      state: TransactionOutputState.State[F],
      blockId: BlockId
  ): F[TransactionOutputState.State[F]] =
    fetchBody(blockId)
      .map(_.transactionIds.reverse)
      .flatMap(
        _.foldLeftM(state)((state, transactionId) =>
          fetchTransaction(transactionId)
            .flatMap(transaction =>
              state.remove(transactionId) *>
                transaction.inputs.reverse.foldLeftM(state)((state, input) =>
                  OptionT
                    .fromOption[F](input.reference.transactionId)
                    .getOrRaise(new IllegalArgumentException("Illegal SelfSpend"))
                    .flatMap(transactionId =>
                      OptionT(state.get(transactionId))
                        .getOrElseF(
                          fetchTransaction(transactionId)
                            .map(_.outputs.length)
                            .map(BitVector.low(_))
                        )
                        .map(_.set(input.reference.index))
                        .flatMap(state.put(transactionId, _))
                        .as(state)
                    )
                )
            )
            .as(state)
        )
      )
      .as(state)

File: dart/blockchain/lib/ledger/transaction_output_state.dart
import 'dart:typed_data';

import 'package:blockchain/codecs.dart';
import 'package:blockchain/common/event_sourced_state.dart';
import 'package:blockchain/common/parent_child_tree.dart';
import 'package:blockchain/common/store.dart';
import 'package:blockchain_protobuf/models/core.pb.dart';

abstract class TransactionOutputState {
  Future<bool> transactionOutputIsSpendable(
      BlockId blockId, TransactionOutputReference outputReference);
}

typedef State = Store<TransactionId, Uint32List>;
typedef FetchBlockBody = Future<BlockBody> Function(BlockId);
typedef FetchTransaction = Future<Transaction> Function(TransactionId);

class TransactionOutputStateImpl extends TransactionOutputState {
  final BlockSourcedState<State> eventSourcedState;

  TransactionOutputStateImpl(this.eventSourcedState);

  @override
  Future<bool> transactionOutputIsSpendable(
      BlockId blockId, TransactionOutputReference outputReference) async {
    final spendableIndices = await eventSourcedState.useStateAt(
        blockId, (state) => state.get(outputReference.transactionId));
    return spendableIndices != null &&
        spendableIndices.contains(outputReference.index);
  }

  factory TransactionOutputStateImpl.make(
      State initialState,
      BlockId currentBlockId,
      Future<BlockBody> Function(BlockId) fetchBlockBody,
      Future<Transaction> Function(TransactionId) fetchTransaction,
      ParentChildTree<BlockId> parentChildTree,
      Future<void> Function(BlockId) currentEventChanged) {
    final eventState = BlockSourcedState<State>(
      (state, blockId) => _applyBlock(
        fetchBlockBody,
        fetchTransaction,
        state,
        blockId,
      ),
      (state, blockId) => _unapplyBlock(
        fetchBlockBody,
        fetchTransaction,
        state,
        blockId,
      ),
      parentChildTree,
      initialState,
      currentBlockId,
      currentEventChanged,
    );
    return TransactionOutputStateImpl(eventState);
  }
}

Future<State> _applyBlock(FetchBlockBody fetchBlockBody,
    FetchTransaction fetchTransaction, State state, BlockId blockId) async {
  final body = await fetchBlockBody(blockId);
  for (final transactionId in body.transactionIds) {
    final transaction = await fetchTransaction(transactionId);
    for (final input in transaction.inputs) {
      final spentTxId = input.reference.transactionId;
      final unspentIndices = await state.getOrRaise(spentTxId);
      final newUnspentIndices = Uint32List.fromList(unspentIndices
          .where((index) => index != input.reference.index)
          .toList());
      if (newUnspentIndices.isEmpty) {
        await state.remove(spentTxId);
      } else {
        state.put(spentTxId, newUnspentIndices);
      }
    }
    if (transaction.outputs.isNotEmpty) {
      final indices = Uint32List.fromList(
          List.generate(transaction.outputs.length, (i) => i));
      await state.put(await transaction.id, indices);
    }
  }
  return state;
}

Future<State> _unapplyBlock(FetchBlockBody fetchBlockBody,
    FetchTransaction fetchTransaction, State state, BlockId blockId) async {
  final body = await fetchBlockBody(blockId);
  for (final transactionId in body.transactionIds) {
    final transaction = await fetchTransaction(transactionId);
    state.remove(transactionId);
    for (final input in transaction.inputs) {
      final spentTxId = input.reference.transactionId;
      final unspentIndices = await state.get(spentTxId);
      if (unspentIndices != null) {
        final newUnspentIndices = Uint32List.fromList(unspentIndices)
          ..add(input.reference.index);
        await state.put(spentTxId, newUnspentIndices);
      } else {
        await state.put(
            spentTxId, Uint32List.fromList([input.reference.index]));
      }
    }
  }
  return state;
}
